shoes-swt/lib/shoes/swt/common/painter.rb

```diff
@@ -30,22 +30,17 @@ def paint_control(event)
         end
 
         def paint_object(graphics_context)
-          cap = LINECAP[@obj.dsl.style[:cap]]
+          dsl = @obj.dsl
+          cap = LINECAP[dsl.style[:cap]]
           graphics_context.set_line_cap(cap) if cap
           graphics_context.set_transform(@obj.transform)
-          obj = @obj.dsl
-          case obj
-            when ::Shoes::Oval, ::Shoes::Rect
-              set_rotate graphics_context, obj.rotate,
-                         obj.element_left + obj.element_width / 2.0,
-                         obj.element_top + obj.element_height / 2.0 do
-                fill graphics_context if fill_setup(graphics_context)
-                draw graphics_context if draw_setup(graphics_context)
-              end
-            else
-              fill graphics_context if fill_setup(graphics_context)
-              draw graphics_context if draw_setup(graphics_context)
-          end
+
+          rotation = rotate(graphics_context, dsl) if dsl.needs_rotate?
+
+          fill graphics_context if fill_setup(graphics_context)
+          draw graphics_context if draw_setup(graphics_context)
+        ensure
+          undo_rotate(graphics_context, dsl, rotation) if rotation
```

Comment:
So before we were calling `set_rotate`, which took a block and did work before and after it. `undo_rotate` in the new code is just what was after the `yield` in `set_rotate` before.

Reading over it again, I'm with you that it's probably be better to just extract the `fill` + `draw`, so unless you object will roll with that.